### PR TITLE
drivers: led: Introduce LP5521 driver

### PR DESCRIPTION
**Pull request by AyushKot96**
### Description
This PR adds a new common driver for controlling [LP5521](https://www.ti.com/lit/ds/symlink/lp5521.pdf?ts=1716362846409&ref_url=https%253A%252F%252Fwww.ti.com%252Fproduct%252FLP5521), supported by CX1825 (in-tree hardware) and [LP5562](https://www.ti.com/lit/ds/symlink/lp5562.pdf?ts=1718290176189&ref_url=https%253A%252F%252Fwww.google.com%252F)

### Changes
- Introduce LP55XX LED Driver to control LP5521 and LP5562.
- Remove old LP5562 driver files.
- Add LP5521 support to existing `croxel_cx1825`.

### Testing
Testing has been done using samples/drivers/led_lp5521 application on croxel_cx1825 board.
``` console
west build -b croxel_cx1825/nrf52840 samples/drivers/led_lp5521
```
Due to unavailability of LP5562, its testing is currently pending.

### CODE REVIEW

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:630]: Let's fix formatting on Function declarations/definitions throughout the whole PR (just capturing it once)

**Review by ubieda — CHANGES_REQUESTED** [review]
First pass review

File: samples/drivers/led_lp5521/README.rst
Comment (ubieda) [inline samples/drivers/led_lp5521/README.rst:18]:
```suggestion
Build the application for the :ref:`croxel_cx1825_nrf52840` board, which includes an LP5521 LED driver.
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:919]: This is not addressed. Please format back-slashes
```suggestion
#define LP5521_DEFINE(id)							\
	BUILD_ASSERT(DT_INST_PROP(id, red_output_current) <=			\
						LP5521_MAX_CURRENT_SETTING,	\
		"Red channel current must be between 0 and 25.5 mA.");		\
	BUILD_ASSERT(DT_INST_PROP(id, green_output_current) <=			\
						LP5521_MAX_CURRENT_SETTING,	\
		"Green channel current must be between 0 and 25.5 mA.");	\
	BUILD_ASSERT(DT_INST_PROP(id, blue_output_current) <=			\
						LP5521_MAX_CURRENT_SETTING,	\
		"Blue channel current must be between 0 and 25.5 mA.");		\
										\
	static const struct lp5521_config lp5521_config_##id = {		\
		.bus = I2C_DT_SPEC_INST_GET(id),				\
		.r_current = DT_INST_PROP(id, red_output_current),		\
		.g_current = DT_INST_PROP(id, green_output_current),		\
		.b_current = DT_INST_PROP(id, blue_output_current),		\
		.en_pin = GPIO_DT_SPEC_INST_GET_OR(id, en_gpios, {0}),		\
	};									\
										\
	struct lp5521_data lp5521_data_##id;					\
	DEVICE_DT_INST_DEFINE(id, &lp5521_led_init, NULL,			\
			&lp5521_data_##id,					\
			&lp5521_config_##id, POST_KERNEL,			\
			CONFIG_LED_INIT_PRIORITY,				\
			&lp5521_led_api);

```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:801]:
```suggestion
	if ((curr_op_mode != LP5521_OP_MODE_DIRECT_CTRL) &&
	    (curr_op_mode != LP5521_OP_MODE_DISABLED)) {
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:681]:
```suggestion
	ret = lp5521_program_set_brightness(dev, led, command_index,
								     dev_data->max_brightness);
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:736]:
```suggestion
	if ((current_mode != LP5521_OP_MODE_DIRECT_CTRL) &&
	    (current_mode != LP5521_OP_MODE_DISABLED)) {
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:727]:
```suggestion
	if ((value < dev_data->min_brightness) ||
	    (value > dev_data->max_brightness)) {
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:868]:
```suggestion
	if (i2c_reg_write_byte_dt(&config->bus, LP5521_ENABLE,
	    LP5521_ENABLE_CHIP_EN)) {
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:638]:
```suggestion
lp5521_update_blinking_brightness(const struct device *dev,
							  enum lp5521_led_channels led,
							  uint8_t brightness_on)
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:607]:
```suggestion
lp5521_program_go_to_start(const struct device *dev,
						enum lp5521_led_channels led,
						uint8_t command_index)
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:580]:
```suggestion
lp5521_program_wait(const struct device *dev,
				    enum lp5521_led_channels led,
				    uint8_t command_index,
				    uint32_t time)
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:545]:
```suggestion
lp5521_program_ramp(const struct device *dev,
				      enum lp5521_led_channels led,
				      uint8_t command_index,
				      uint32_t time_per_step,
				      uint8_t step_count,
				      enum lp5521_engine_fade_dirs fade_dir)
```

File: drivers/led/lp5521.c
Comment (ubieda) [inline drivers/led/lp5521.c:505]:
```suggestion
lp5521_program_set_brightness(const struct device *dev,
						     enum lp5521_led_channels led,
						     uint8_t command_index,
						     uint8_t brightness)
```

**Review by ubieda — CHANGES_REQUESTED** [review]
Formatting is off on `lp5521.c` I pointed out a few but please fix it altogether.

**Comment by simonguinot** [comment]
Hi @AyushKot96,

Thanks for this new PR :)

Can it be merged with the `lp55{62,69}` drivers ? I only looked for a few seconds but I saw similarities with the registers addresses. Please check.

I'll try to review this PR later this week.

**Comment by AyushKot96** [comment]
> Hi @AyushKot96,
> 
> Thanks for this new PR :)
> 
> Can it be merged with the `lp5562` driver ? I only looked for a few seconds but I saw similarities with the registers addresses. Please check.
> 
> I'll try to review this PR later this week.

Both IC's are similar but many registers are different. LP5562 driver wont work with LP5521.

**Review by ubieda — DISMISSED** [review]

**Review by thedjnK — DISMISSED** [review]

**Review by simonguinot — CHANGES_REQUESTED** [review]
Hi @AyushKot96,

Thanks for contributing this driver !

I started my review but the more code I read, the more I find it very similar to the `lp5562` driver. So I stop here and I reiterate my initial question: what would prevent to have a common driver for the `lp5521` and `lp5562` chips ? I understand that there are slight differences with some registers but they can be abstracted. Please comment.

**Comment by simonguinot** [comment]
Hi @JoeHut, @Jamal-Ra-Davis, @RickBruyninckx what do you think of the idea of merging the tp5521 and lp5562 drivers ?

**Comment by ubieda** [comment]
Friendly reminder: Feature Freeze for the next Release is this upcoming friday.
Do you believe this PR can be targetted to get in by then, @AyushKot96 @simonguinot?

**Comment by RickBruyninckx** [comment]
@simonguinot while we're at it, can't we just make a lp55xx driver covering all of them? (I will look Monday how much the lp55xx family shares functionality)

**Comment by simonguinot** [comment]
> Friendly reminder: Feature Freeze for the next Release is this upcoming friday. Do you believe this PR can be targetted to get in by then, @AyushKot96 @simonguinot?

@ubieda I think it's unlikely. For now my opinion is that some of the "lp55xx" drivers should be merged. So if I'm correct, the "lp5521" driver cannot be merged into its current form. And again I won't rush it.

**Comment by RickBruyninckx** [comment]
> > @simonguinot while we're at it, can't we just make a lp55xx driver covering all of them? (I will look Monday how much the lp55xx family shares functionality)
> 
> Thanks @RickBruyninckx. Let me know !

I glanced at the drivers made for linux and documented at TI and they have a seperate driver for each of the leddrivers but with a "leds-lp55xx-common.h" file for the generic lp55xx stuff and a common devicetree binding:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/leds/leds-lp55xx.yaml

We can do something similar right?

**Comment by ubieda** [comment]
> > Friendly reminder: Feature Freeze for the next Release is this upcoming friday. Do you believe this PR can be targetted to get in by then, @AyushKot96 @simonguinot?
> 
> @ubieda I think it's unlikely. For now my opinion is that some of the `lp55xx` drivers should be merged. So if I'm correct, the `lp5521` driver cannot be merged into its current form. And again I won't rush it.

There's a clear upside in refactoring these drivers into common code, as it makes it easier to maintain and catch bugs. As pointed out by @RickBruyninckx, this has been somewhat done on Linux so it's possible.

However, I disagree with this being a requirement for merging this PR, for the following reasons:
- First and foremost: This feature is important for us (Croxel) as it provides upstream support for a feature on our board (https://docs.zephyrproject.org/latest/boards/croxel/croxel_cx1825/doc/index.html) and we would highly benefit from it making it to the release.
- The review process of a PR introducting a new driver and the refactoring of existing functional drivers into common code are very different from each other.
- This suggested refactoring is not implemented for the existing LP55XX instances (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/lp5562.c and https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/lp5569.c) so this PR is not breaking a pattern, nor ignoring an already filed issue. Because of this: this refactoring efforts naturally falls into a GH enhancement issue (which we can happily capture before merging this PR).

Long-story short: the benefits for getting this feature for 3.7.0 far outweigh the cost of introducing it as a separate driver now and refactoring it later.

@simonguinot @RickBruyninckx please, reconsider your stance on this.

**Comment by ubieda** [comment]
@AyushKot96 please fix CI failures. Rememeber you can run these locally:

``` console
twister -p croxel_cx1825/nrf52840 -T samples
```

**Comment by ubieda** [comment]
> Because of this: this refactoring efforts naturally falls into a GH enhancement issue (which we can happily capture before merging this PR).

Issue captured: #74027

**Comment by simonguinot** [comment]
> > > Friendly reminder: Feature Freeze for the next Release is this upcoming friday. Do you believe this PR can be targetted to get in by then, @AyushKot96 @simonguinot?
> > 
> > 
> > @ubieda I think it's unlikely. For now my opinion is that some of the `lp55xx` drivers should be merged. So if I'm correct, the `lp5521` driver cannot be merged into its current form. And again I won't rush it.
> 
> There's a clear upside in refactoring these drivers into common code, as it makes it easier to maintain and catch bugs. As pointed out by @RickBruyninckx, this has been somewhat done on Linux so it's possible.
> 
> However, I disagree with this being a requirement for merging this PR, for the following reasons:
> 
>     * First and foremost: This feature is important for us (Croxel) as it provides upstream support for a feature on our board (https://docs.zephyrproject.org/latest/boards/croxel/croxel_cx1825/doc/index.html) and we would highly benefit from it making it to the release.
> 
>     * The review process of a PR introducting a new driver and the refactoring of existing functional drivers into common code are very different from each other.
> 
>     * This suggested refactoring is not implemented for the existing LP55XX instances (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/lp5562.c and https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/lp5569.c) so this PR is not breaking a pattern, nor ignoring an already filed issue. Because of this: this refactoring efforts naturally falls into a GH enhancement issue (which we can happily capture before merging this PR).
> 
> 
> Long-story short: the benefits for getting this feature for 3.7.0 far outweigh the cost of introducing it as a separate driver now and refactoring it later.
> 
> @simonguinot @RickBruyninckx please, reconsider your stance on this.

@ubieda I understand this feature is important for your company.

But here is my problem. I can see a lot of similarities between the lp5521 and lp5562 drivers. And I am opposed to the addition of a new driver duplicating code with an existing one if the latter can be extended. So I am trying to evaluate the cost of a such extension here and I am asking @AyushKot96 about it. I think it is a legitimate request which should have been anticipated considering that the code of the lp5521 driver has been copied from lp5562. Please note that I already raised this point two weeks ago and here is the answer I got:

> Both IC's are similar but many register addresses are different. LP5562 driver wont work with LP5521.

About the registers addresses it is not accurate. I can see that _most_ of them are identical. And I already know that the lp5562 driver won't work with the LP5521 controller in its current state. But that didn't answer my question...

@AyushKot96 is that possible to add support for the LP5521 controller to the lp5562 driver ? have you considered this option ? And why did choose to write a separate driver ?

**Comment by AyushKot96** [comment]
> > @simonguinot @RickBruyninckx please, reconsider your stance on this.
> 
> @ubieda I understand this feature is important for your company.
> 
> But here is my problem. I can see a lot of similarities between the lp5521 and lp5562 drivers. And I am opposed to the addition of a new driver duplicating code with an existing one if the latter can be extended. So I am trying to evaluate the cost of a such extension here and I am asking @AyushKot96 about it. I think it is a legitimate request which should have been anticipated considering that the code of the lp5521 driver has been copied from lp5562. Please note that I already raised this point two weeks ago and here is the answer I got:
> 
> > Both IC's are similar but many register addresses are different. LP5562 driver wont work with LP5521.
> 
> About the registers addresses it is not accurate. I can see that **most** of them are identical. And I already know that the lp5562 driver won't work with the LP5521 controller in its current state. But that didn't answer my question...
> 
> @AyushKot96 is that possible to add support for the LP5521 controller to the lp5562 driver ? have you considered this option ? And why did choose to write a separate driver ?

@simonguinot Upon revisiting the drivers, I came to know that I could have refactor them and make a common driver. I am prepared to address this in a separate pull request.

**Comment by bbilas** [comment]
> Hi @AyushKot96,
> 
> Thanks for contributing this driver !
> 
> I started my review but the more code I read, the more I find it very similar to the `lp5562` driver. So I stop here and I reiterate my initial question: what would prevent to have a common driver for the `lp5521` and `lp5562` chips ? I understand that there are slight differences with some registers but they can be abstracted. Please comment.

I agree with Simon. Please create a generic `lp55xx` driver that will support both variants. The current solution is unacceptable to me as there is no need to have another driver which was mostly copied from the other one.

**Comment by kartben** [comment]
> @simonguinot Upon revisiting the drivers, I came to know that I could have refactor them and make a common driver. I am prepared to address this in a separate pull request.

Hi @AyushKot96 - I think it's fair that the maintainer of a particular area would not want to accept new code that's duplicated if it's clear that there is a better path. That's how technical debt can quickly add up in an open source project as the promised follow-up PR may or may not ever show up (and I am by no means implying that you, in particular, won't follow up, but you get the idea).

@ubieda 

> However, I disagree with this being a requirement for merging this PR, for the following reasons:
> 
> First and foremost: This feature is important for us (Croxel) as it provides upstream support for a feature on our board ([docs.zephyrproject.org/latest/boards/croxel/croxel_cx1825/doc/index.html](https://docs.zephyrproject.org/latest/boards/croxel/croxel_cx1825/doc/index.html)) and we would highly benefit from it making it to the release.

Of course! On the other hand it's also worth pointing out it's a not unsignificant amount of code that was submitted 3-ish weeks ahead of feature freeze :) And the feedback re: trying to identify common code was mentioned very early in the review process.

> The review process of a PR introducting a new driver and the refactoring of existing functional drivers into common code are very different from each other.
> This suggested refactoring is not implemented for the existing LP55XX instances ([main/drivers/led/lp5562.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/lp5562.c?rgh-link-date=2024-06-10T16%3A22%3A22Z) and [main/drivers/led/lp5569.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/lp5569.c?rgh-link-date=2024-06-10T16%3A22%3A22Z)) so this PR is not breaking a pattern, nor ignoring an already filed issue. Because of this: this refactoring efforts naturally falls into a GH enhancement issue (which we can happily capture before merging this PR).

It seems to me as the lp5569 driver is *vastly* different from the existing LP5562, where the LP5512 one is effectively copy-pasting *a lot* of existing code?

**Comment by ubieda** [comment]
@simonguinot @kartben Understood. We'll refactor the driver. Thanks!
CC: @AyushKot96

File: boards/croxel/croxel_cx1825/croxel_cx1825_nrf52840.dts
Comment (ubieda) [inline boards/croxel/croxel_cx1825/croxel_cx1825_nrf52840.dts:84]: Needs rebasing. This has been updated already: #74893

**Comment by ubieda** [comment]
@simonguinot Seems like the refactoring has been done. Would you mind taking another look at this? Thanks

**Comment by simonguinot** [comment]
> @simonguinot Seems like the refactoring has been done. Would you mind taking another look at this? Thanks

Hello @ubieda,

I'll look at it next week.

Note that I already reviewed https://github.com/zephyrproject-rtos/zephyr/pull/72713 from @AyushKot96 last month but I didn't get any answer since...  That's why I was hesitant to start the review for this PR.

**Comment by ubieda** [comment]
> > @simonguinot Seems like the refactoring has been done. Would you mind taking another look at this? Thanks
> 
> Hello @ubieda,
> 
> I'll look at it next week.
> 
> Note that I already reviewed #72713 from @AyushKot96 last month but I didn't get any answer since... That's why I was hesitant to start the review for this PR.

Thanks! We'll come back to that one soon (it's in our backlog). CC: @cx-anuj-pathak

**Review by simonguinot — CHANGES_REQUESTED** [review]
Hi @AyushKot96,

You need to help me a little here. I can't review a such big commit with everything in it, including renamed files. You need to structure the pull request with several commits. For example:

- rename the `lp5562` driver and sample into `lp55xx`. So I'll be able to use git to visualize the changes coming in the next commits.
- make the driver and sample generic to prepare for adding support of the LP5521 controller. Don't be afraid to split this step into several commits as well.
- add support for the LP5521 controller to the driver.
- add support for the LP5521 controller to the sample.

Thanks in advance !

**Comment by cx-anuj-pathak** [comment]
@simonguinot we have rearranged commits. but it went upto 14 commits, and in the process we have identified few other things too and a bug in lp5562 driver. So we have decided internally to break it down in 3-4 PR's.

- add en gpio dts property and handling, PR filed https://github.com/zephyrproject-rtos/zephyr/pull/77375
- fix bug in the way blink is handled, if a engine is already linked, it is not reused in subsequent call to blink, instead a new IDLE engine is found and linked. Thus if 3 colors are in blink state, we can not change blink period without led_on/off call. Will file PR soon.
- Simplify and Refactor LP5562 code in generic format and add support for LP5521.

Let me know if you have any suggestions? 
We'll close this PR

**Comment by biii-anuj-pathak** [comment]
FIled a new PR https://github.com/zephyrproject-rtos/zephyr/pull/78968

**Pull request closed without merging**
